cranelift/jit/src/memory.rs

```diff
@@ -242,6 +242,51 @@ impl Memory {
         self.already_protected = self.allocations.len();
     }
 
+    /// Prepares an icache flush.
+    ///
+    /// [Memory::icache_flush] *must* be called after this.
+    pub(crate) fn prepare_icache_flush(&self) {
+        #[cfg(all(target_arch = "aarch64", target_os = "linux"))]
+        {
+            let cmd: libc::c_int = 64; // MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED_SYNC_CORE
+
+            // This is a requirement of the membarrier() call executed by
+            // the finalize_definitions() method.
+            unsafe { libc::syscall(libc::SYS_membarrier, cmd) };
+        }
+    }
+
+    /// Flushes the icache to ensure that no processor has a stale view of this memory.
+    ///
+    /// In order to call this function [Memory::prepare_icache_flush] must have been called previously.
+    pub(crate) fn icache_flush(&self) {
```

Comment:
Maybe call this from set_readonly_and_executable? That will also allow skipping parts that have already been flushed previously when doing multiple rounds of defining a function and finalizing everything, like for hot reloading.

Reply:
I've done that and I've also switched the `prepare_icache_flush` calls to be done in `allocate()`. Calling one inside `Memory` and the other outside I think would be a kind of weird API quirk. I added a `MemoryUse` flag to ensure that we always call the icache flushes/prepares in the appropriate place, and only for memory intended for execution.

Another thing that I noticed and think is probably a bug, is that we only call `MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED_SYNC_CORE` once, when initially allocating the memory, but then can call `MEMBARRIER_CMD_PRIVATE_EXPEDITED_SYNC_CORE` multiple times when hotswapping, I'm not totally sure this is allowed and what happens if we do this. 

If it is, then maybe we want to call `prepare_icache_flush` in `new` instead of  `allocate`.

I tested this patch on `aarch64-linux` and `aarch64-msvc` and it seems to work on both.

Reply:
Refer to [my comment](https://github.com/bytecodealliance/wasmtime/pull/4987#discussion_r984806226) in #4987, but the gist of it is that `MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED_SYNC_CORE` should be executed only once for the whole lifetime of the process and subsequent calls are just unnecessary overhead, yet harmless. So, yes, you shouldn't have moved the code.

BTW another subtle issue - the `membarrier()` calls are completely unnecessary for single-threaded processes and multi-threaded programs, provided that each thread uses its own generated code. However, I can't think of any easy way to distinguish these use cases from the regular multi-threaded one, and I am not sure if adding an option to `JITModule::new()` is the right approach.

Reply:
Got it! I'll move it to `new` and add a comment with that description. Thanks!

Reply:
Got it! I'd like to try and keep all the `membarrier` calls contained to the `Memory` struct, otherwise I think we get a really weird interface where you need to do some stuff outside of `Memory` and some other stuff happens automatically. I think moving this to `new` effectively gets us back to where we were before while keeping these details contained.

I'm not really seeing a way of cleaning the issue you mentioned in https://github.com/bytecodealliance/wasmtime/pull/4987#discussion_r984806226 where we only really need to do this once per process.